File: backend/app/api/v1/endpoints/reports.py
"""Management analytics/reporting (M11). Every route resolves the
caller's authorized store list via
app.modules.reports.service.resolve_authorized_store_ids BEFORE calling
any service function — a store id supplied by the client can never
bypass authorization (docs/M11_DESIGN.md Section 3/12). Permissions
reuse the existing read permission for each domain (no new permission
constants): `reports.read` for sales/KPI, `accounting.read` for
financial reports, `inventory.read` for inventory analytics,
`purchasing.read` for purchasing/supplier analytics, `payroll.read` for
labor/payroll analytics — matching exactly what already gates reading
the underlying raw data in each domain.
"""

from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.modules.accounting import service as accounting_service
from app.modules.auth.permissions import (
    ACCOUNTING_READ,
    INVENTORY_READ,
    PAYROLL_READ,
    PURCHASING_READ,
    REPORTS_READ,
)
from app.modules.auth.service import CurrentUser, require_permission
from app.modules.reports import schemas as report_schemas
from app.modules.reports import service as reports_service

router = APIRouter(prefix="/reports", tags=["reports"])

_reports_permission = require_permission(REPORTS_READ)
_accounting_permission = require_permission(ACCOUNTING_READ)
_inventory_permission = require_permission(INVENTORY_READ)
_purchasing_permission = require_permission(PURCHASING_READ)
_payroll_permission = require_permission(PAYROLL_READ)


def _resolve(current_user: CurrentUser, store_id: list[int] | None) -> list[int] | None:
    return reports_service.resolve_authorized_store_ids(current_user, store_id)


# --- Sales & profitability (Phase 3) ----------------------------------------


@router.get("/sales/summary", response_model=report_schemas.SalesSummaryRead)
def get_sales_summary(
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_reports_permission),
) -> report_schemas.SalesSummaryRead:
    resolved = _resolve(current_user, store_id)
    result = reports_service.sales_summary(
        db, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return report_schemas.SalesSummaryRead.model_validate(result)


@router.get("/sales/breakdown", response_model=list[report_schemas.SalesByDimensionRowRead])
def get_sales_breakdown(
    dimension: str = Query(..., pattern="^(store|product)$"),
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_reports_permission),
) -> list[report_schemas.SalesByDimensionRowRead]:
    resolved = _resolve(current_user, store_id)
    fn = (
        reports_service.sales_by_store if dimension == "store" else reports_service.sales_by_product
    )
    rows = fn(db, store_ids=resolved, date_from=date_from, date_to=date_to)
    return [report_schemas.SalesByDimensionRowRead.model_validate(r) for r in rows]


@router.get("/sales/by-payment-method", response_model=list[report_schemas.PaymentMethodRowRead])
def get_sales_by_payment_method(
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_reports_permission),
) -> list[report_schemas.PaymentMethodRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.sales_by_payment_method(
        db, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return [report_schemas.PaymentMethodRowRead.model_validate(r) for r in rows]


@router.get("/sales/trend", response_model=list[report_schemas.SalesTrendPointRead])
def get_sales_trend(
    date_from: date,
    date_to: date,
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_reports_permission),
) -> list[report_schemas.SalesTrendPointRead]:
    resolved = _resolve(current_user, store_id)
    points = reports_service.sales_trend_by_day(
        db, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return [report_schemas.SalesTrendPointRead.model_validate(p) for p in points]


# --- Financial reporting (Phase 4) ------------------------------------------


@router.get("/financial/trial-balance", response_model=list[report_schemas.TrialBalanceRowRead])
def get_financial_trial_balance(
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_accounting_permission),
) -> list[report_schemas.TrialBalanceRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = accounting_service.trial_balance(
        db, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return [report_schemas.TrialBalanceRowRead.model_validate(r) for r in rows]


@router.get("/financial/profit-loss", response_model=report_schemas.ProfitAndLossRead)
def get_financial_profit_loss(
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_accounting_permission),
) -> report_schemas.ProfitAndLossRead:
    resolved = _resolve(current_user, store_id)
    result = accounting_service.profit_and_loss(
        db, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return report_schemas.ProfitAndLossRead.model_validate(result)


@router.get(
    "/financial/profit-loss-comparative", response_model=report_schemas.ProfitAndLossComparativeRead
)
def get_financial_profit_loss_comparative(
    date_from: date,
    date_to: date,
    prior_date_from: date,
    prior_date_to: date,
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_accounting_permission),
) -> report_schemas.ProfitAndLossComparativeRead:
    resolved = _resolve(current_user, store_id)
    result = reports_service.profit_and_loss_comparative(
        db,
        store_ids=resolved,
        date_from=date_from,
        date_to=date_to,
        prior_date_from=prior_date_from,
        prior_date_to=prior_date_to,
    )
    return report_schemas.ProfitAndLossComparativeRead(
        current=report_schemas.ProfitAndLossRead.model_validate(result.current),
        prior=report_schemas.ProfitAndLossRead.model_validate(result.prior),
    )


@router.get("/financial/account-summary", response_model=report_schemas.AccountTypeSummaryRead)
def get_financial_account_summary(
    account_type: str = Query(..., pattern="^(REVENUE|EXPENSE|ASSET|LIABILITY)$"),
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_accounting_permission),
) -> report_schemas.AccountTypeSummaryRead:
    resolved = _resolve(current_user, store_id)
    result = reports_service.account_type_summary(
        db, account_type=account_type, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return report_schemas.AccountTypeSummaryRead(
        account_type=result.account_type,
        rows=[report_schemas.TrialBalanceRowRead.model_validate(r) for r in result.rows],
        total_debit=result.total_debit,
        total_credit=result.total_credit,
    )


@router.get("/financial/balance-sheet", response_model=report_schemas.BalanceSheetSummaryRead)
def get_financial_balance_sheet(
    store_id: list[int] | None = Query(None),
    as_of: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_accounting_permission),
) -> report_schemas.BalanceSheetSummaryRead:
    resolved = _resolve(current_user, store_id)
    result = reports_service.balance_sheet_summary(db, store_ids=resolved, as_of=as_of)
    return report_schemas.BalanceSheetSummaryRead(
        store_ids=result.store_ids,
        as_of=result.as_of,
        assets=[report_schemas.TrialBalanceRowRead.model_validate(r) for r in result.assets],
        liabilities=[
            report_schemas.TrialBalanceRowRead.model_validate(r) for r in result.liabilities
        ],
        total_assets=result.total_assets,
        total_liabilities=result.total_liabilities,
    )


@router.get(
    "/financial/cash-summary",
    response_model=list[report_schemas.PaymentMethodReconciliationRowRead],
)
def get_financial_cash_summary(
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_accounting_permission),
) -> list[report_schemas.PaymentMethodReconciliationRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.cash_payment_method_summary(
        db, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return [report_schemas.PaymentMethodReconciliationRowRead.model_validate(r) for r in rows]


# --- Inventory analytics (Phase 5) ------------------------------------------


@router.get("/inventory/value", response_model=list[report_schemas.InventoryValueRowRead])
def get_inventory_value(
    group_by: str = Query("store", pattern="^(store|category)$"),
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_inventory_permission),
) -> list[report_schemas.InventoryValueRowRead]:
    resolved = _resolve(current_user, store_id)
    fn = (
        reports_service.inventory_value_by_store
        if group_by == "store"
        else reports_service.inventory_value_by_category
    )
    rows = fn(db, store_ids=resolved)
    return [report_schemas.InventoryValueRowRead.model_validate(r) for r in rows]


@router.get("/inventory/movements", response_model=list[report_schemas.MovementSummaryRowRead])
def get_inventory_movements(
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_inventory_permission),
) -> list[report_schemas.MovementSummaryRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.inventory_movement_summary(
        db, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return [report_schemas.MovementSummaryRowRead.model_validate(r) for r in rows]


@router.get("/inventory/shrinkage", response_model=list[report_schemas.ShrinkageRowRead])
def get_inventory_shrinkage(
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_inventory_permission),
) -> list[report_schemas.ShrinkageRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.shrinkage_summary(
        db, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return [report_schemas.ShrinkageRowRead.model_validate(r) for r in rows]


@router.get("/inventory/turnover", response_model=list[report_schemas.TurnoverRowRead])
def get_inventory_turnover(
    date_from: date,
    date_to: date,
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_inventory_permission),
) -> list[report_schemas.TurnoverRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.inventory_turnover(
        db, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return [report_schemas.TurnoverRowRead.model_validate(r) for r in rows]


@router.get("/inventory/slow-moving", response_model=list[report_schemas.SlowMovingRowRead])
def get_inventory_slow_moving(
    store_id: list[int] | None = Query(None),
    window_days: int = 90,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_inventory_permission),
) -> list[report_schemas.SlowMovingRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.slow_moving_products(db, store_ids=resolved, window_days=window_days)
    return [report_schemas.SlowMovingRowRead.model_validate(r) for r in rows]


@router.get("/inventory/stockouts", response_model=list[report_schemas.StockoutRowRead])
def get_inventory_stockouts(
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_inventory_permission),
) -> list[report_schemas.StockoutRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.stockouts(db, store_ids=resolved)
    return [report_schemas.StockoutRowRead.model_validate(r) for r in rows]


@router.get("/inventory/negative-stock", response_model=list[report_schemas.NegativeStockRowRead])
def get_inventory_negative_stock(
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_inventory_permission),
) -> list[report_schemas.NegativeStockRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.negative_stock_products(db, store_ids=resolved)
    return [report_schemas.NegativeStockRowRead.model_validate(r) for r in rows]


@router.get("/inventory/in-transit", response_model=list[report_schemas.InTransitRowRead])
def get_inventory_in_transit(
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_inventory_permission),
) -> list[report_schemas.InTransitRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.inventory_in_transit(db, store_ids=resolved)
    return [report_schemas.InTransitRowRead.model_validate(r) for r in rows]


@router.get(
    "/inventory/in-transit-reconciliation", response_model=report_schemas.GlReconciliationRowRead
)
def get_inventory_in_transit_reconciliation(
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_inventory_permission),
) -> report_schemas.GlReconciliationRowRead:
    # Deliberately no store filter: matches the existing M8 endpoint
    # (GET /transfers/reports/inventory-in-transit-reconciliation) this
    # wraps, which is company-wide by design -- the account represents
    # value in transit BETWEEN stores, not within one.
    result = reports_service.in_transit_reconciliation(db)
    return report_schemas.GlReconciliationRowRead.model_validate(result)


@router.get(
    "/inventory/stock-counts/{stock_count_id}/variance",
    response_model=list[report_schemas.StockCountVarianceRowRead],
)
def get_stock_count_variance(
    stock_count_id: int,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_inventory_permission),
) -> list[report_schemas.StockCountVarianceRowRead]:
    from app.core.exceptions import ForbiddenError
    from app.modules.inventory.models import StockCount

    count = db.get(StockCount, stock_count_id)
    if (
        count is not None
        and current_user.store_id is not None
        and current_user.store_id != count.store_id
    ):
        raise ForbiddenError(
            f"Your account is scoped to store {current_user.store_id} and cannot view "
            f"stock count {stock_count_id} in store {count.store_id}",
            error_code="STORE_ACCESS_DENIED",
        )
    rows = reports_service.stock_count_variance(db, stock_count_id=stock_count_id)
    return [report_schemas.StockCountVarianceRowRead.model_validate(r) for r in rows]


# --- Purchasing & supplier analytics (Phase 6) ------------------------------


@router.get("/purchasing/spend", response_model=list[report_schemas.PurchaseSpendRowRead])
def get_purchasing_spend(
    group_by: str = Query("supplier", pattern="^(supplier|store|product)$"),
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_purchasing_permission),
) -> list[report_schemas.PurchaseSpendRowRead]:
    resolved = _resolve(current_user, store_id)
    fn = {
        "supplier": reports_service.purchase_spend_by_supplier,
        "store": reports_service.purchase_spend_by_store,
        "product": reports_service.purchase_spend_by_product,
    }[group_by]
    rows = fn(db, store_ids=resolved, date_from=date_from, date_to=date_to)
    return [report_schemas.PurchaseSpendRowRead.model_validate(r) for r in rows]


@router.get("/purchasing/po-fulfillment", response_model=list[report_schemas.PoFulfillmentRowRead])
def get_purchasing_po_fulfillment(
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_purchasing_permission),
) -> list[report_schemas.PoFulfillmentRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.po_fulfillment(db, store_ids=resolved)
    return [report_schemas.PoFulfillmentRowRead.model_validate(r) for r in rows]


@router.get(
    "/purchasing/delivery-performance",
    response_model=list[report_schemas.SupplierDeliveryPerformanceRowRead],
)
def get_purchasing_delivery_performance(
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_purchasing_permission),
) -> list[report_schemas.SupplierDeliveryPerformanceRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.supplier_delivery_performance(db, store_ids=resolved)
    return [report_schemas.SupplierDeliveryPerformanceRowRead.model_validate(r) for r in rows]


@router.get(
    "/purchasing/price-variance", response_model=list[report_schemas.PurchasePriceVarianceRowRead]
)
def get_purchasing_price_variance(
    store_id: list[int] | None = Query(None),
    date_from: date | None = None,
    date_to: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_purchasing_permission),
) -> list[report_schemas.PurchasePriceVarianceRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.purchase_price_variance_report(
        db, store_ids=resolved, date_from=date_from, date_to=date_to
    )
    return [report_schemas.PurchasePriceVarianceRowRead.model_validate(r) for r in rows]


# --- Labor & payroll analytics (Phase 7) ------------------------------------


@router.get("/payroll/headcount", response_model=list[report_schemas.HeadcountRowRead])
def get_payroll_headcount(
    store_id: list[int] | None = Query(None),
    as_of: date | None = None,
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_payroll_permission),
) -> list[report_schemas.HeadcountRowRead]:
    resolved = _resolve(current_user, store_id)
    rows = reports_service.headcount_by_store(db, store_ids=resolved, as_of=as_of)
    return [report_schemas.HeadcountRowRead.model_validate(r) for r in rows]


@router.get("/payroll/cost-summary", response_model=report_schemas.PayrollCostSummaryRead)
def get_payroll_cost_summary(
    period_start: date,
    period_end: date,
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_payroll_permission),
) -> report_schemas.PayrollCostSummaryRead:
    resolved = _resolve(current_user, store_id)
    result = reports_service.payroll_cost_summary(
        db, store_ids=resolved, period_start=period_start, period_end=period_end
    )
    return report_schemas.PayrollCostSummaryRead.model_validate(result)


@router.get("/payroll/labor-cost-percent", response_model=report_schemas.LaborCostPercentRowRead)
def get_payroll_labor_cost_percent(
    period_start: date,
    period_end: date,
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_payroll_permission),
) -> report_schemas.LaborCostPercentRowRead:
    resolved = _resolve(current_user, store_id)
    result = reports_service.labor_cost_percent_of_sales(
        db, store_ids=resolved, period_start=period_start, period_end=period_end
    )
    return report_schemas.LaborCostPercentRowRead.model_validate(result)


@router.get("/payroll/reconciliation", response_model=report_schemas.GlReconciliationRowRead)
def get_payroll_reconciliation(
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_payroll_permission),
) -> report_schemas.GlReconciliationRowRead:
    resolved = _resolve(current_user, store_id)
    result = reports_service.payroll_gl_reconciliation(db, store_ids=resolved)
    return report_schemas.GlReconciliationRowRead.model_validate(result)


# --- KPI dashboard (Phase 8) -------------------------------------------------


@router.get("/dashboard", response_model=report_schemas.KpiDashboardRead)
def get_kpi_dashboard(
    period_start: date,
    period_end: date,
    store_id: list[int] | None = Query(None),
    db: Session = Depends(get_db),
    current_user: CurrentUser = Depends(_reports_permission),
) -> report_schemas.KpiDashboardRead:
    resolved = _resolve(current_user, store_id)
    result = reports_service.kpi_dashboard(
        db, store_ids=resolved, period_start=period_start, period_end=period_end
    )
    return report_schemas.KpiDashboardRead.model_validate(result)

File: frontend/src/api/reports.ts
/**
 * M16: read-only client for the M11 reporting API
 * (app/api/v1/endpoints/reports.py). Every interface mirrors a Pydantic
 * schema in app/modules/reports/schemas.py exactly — no field is added,
 * renamed, or recomputed here. All monetary/quantity fields stay
 * strings (server-formatted Decimals) and must never be added/multiplied
 * as JS numbers except for display-only aggregation the server itself
 * already did.
 */
import { apiFetch, buildQuery } from './client'

export interface ReportDateRange {
  store_id?: number[]
  date_from?: string
  date_to?: string
}

// --- Sales & profitability ---------------------------------------------

export interface SalesSummary {
  store_ids: number[] | null
  date_from: string | null
  date_to: string | null
  gross_sales: string
  discounts: string
  returns: string
  void_count: number
  void_amount: string
  net_sales: string
  tax: string
  cogs: string
  gross_profit: string
  gross_margin_percent: string | null
  transaction_count: number
  units_sold: string
  average_transaction_value: string | null
}

export function getSalesSummary(params: ReportDateRange = {}): Promise<SalesSummary> {
  return apiFetch<SalesSummary>(`/api/v1/reports/sales/summary${buildQuery(params)}`)
}

export interface SalesByDimensionRow {
  key: number | string
  label: string
  gross_sales: string
  returns: string
  net_sales: string
  cogs: string
  units_sold: string
  transaction_count: number
}

export function getSalesBreakdown(
  dimension: 'store' | 'product',
  params: ReportDateRange = {},
): Promise<SalesByDimensionRow[]> {
  return apiFetch<SalesByDimensionRow[]>(
    `/api/v1/reports/sales/breakdown${buildQuery({ dimension, ...params })}`,
  )
}

export interface PaymentMethodRow {
  payment_method: string
  amount: string
  payment_count: number
}

export function getSalesByPaymentMethod(params: ReportDateRange = {}): Promise<PaymentMethodRow[]> {
  return apiFetch<PaymentMethodRow[]>(
    `/api/v1/reports/sales/by-payment-method${buildQuery(params)}`,
  )
}

export interface SalesTrendPoint {
  sale_date: string
  gross_sales: string
  returns: string
  net_sales: string
  transaction_count: number
}

export function getSalesTrend(
  date_from: string,
  date_to: string,
  store_id?: number[],
): Promise<SalesTrendPoint[]> {
  return apiFetch<SalesTrendPoint[]>(
    `/api/v1/reports/sales/trend${buildQuery({ date_from, date_to, store_id })}`,
  )
}

// --- Financial (GL-derived) ---------------------------------------------

export interface TrialBalanceRow {
  account_code: string
  account_name: string
  account_type: string
  normal_balance: string
  total_debit: string
  total_credit: string
}

export function getTrialBalance(params: ReportDateRange = {}): Promise<TrialBalanceRow[]> {
  return apiFetch<TrialBalanceRow[]>(`/api/v1/reports/financial/trial-balance${buildQuery(params)}`)
}

export interface ProfitAndLoss {
  store_id: number | null
  date_from: string | null
  date_to: string | null
  net_sales: string
  cogs: string
  gross_profit: string
  other_income: string
  operating_expenses: string
  net_income: string
}

export function getProfitAndLoss(params: ReportDateRange = {}): Promise<ProfitAndLoss> {
  return apiFetch<ProfitAndLoss>(`/api/v1/reports/financial/profit-loss${buildQuery(params)}`)
}

export interface ProfitAndLossComparative {
  current: ProfitAndLoss
  prior: ProfitAndLoss
}

export function getProfitAndLossComparative(params: {
  date_from: string
  date_to: string
  prior_date_from: string
  prior_date_to: string
  store_id?: number[]
}): Promise<ProfitAndLossComparative> {
  return apiFetch<ProfitAndLossComparative>(
    `/api/v1/reports/financial/profit-loss-comparative${buildQuery(params)}`,
  )
}

export interface AccountTypeSummary {
  account_type: string
  rows: TrialBalanceRow[]
  total_debit: string
  total_credit: string
}

export function getAccountSummary(
  account_type: 'REVENUE' | 'EXPENSE' | 'ASSET' | 'LIABILITY',
  params: ReportDateRange = {},
): Promise<AccountTypeSummary> {
  return apiFetch<AccountTypeSummary>(
    `/api/v1/reports/financial/account-summary${buildQuery({ account_type, ...params })}`,
  )
}

export interface BalanceSheetSummary {
  store_ids: number[] | null
  as_of: string | null
  assets: TrialBalanceRow[]
  liabilities: TrialBalanceRow[]
  total_assets: string
  total_liabilities: string
}

export function getBalanceSheet(params: {
  store_id?: number[]
  as_of?: string
}): Promise<BalanceSheetSummary> {
  return apiFetch<BalanceSheetSummary>(
    `/api/v1/reports/financial/balance-sheet${buildQuery(params)}`,
  )
}

export interface PaymentMethodReconciliationRow {
  payment_method: string
  operational_amount: string
  gl_account_code: string
  gl_balance: string
  discrepancy: string
}

export function getCashSummary(
  params: ReportDateRange = {},
): Promise<PaymentMethodReconciliationRow[]> {
  return apiFetch<PaymentMethodReconciliationRow[]>(
    `/api/v1/reports/financial/cash-summary${buildQuery(params)}`,
  )
}

// --- Inventory analytics --------------------------------------------------

export interface InventoryValueRow {
  key: number
  label: string
  quantity_on_hand: string
  value: string
}

export function getInventoryValue(
  group_by: 'store' | 'category',
  store_id?: number[],
): Promise<InventoryValueRow[]> {
  return apiFetch<InventoryValueRow[]>(
    `/api/v1/reports/inventory/value${buildQuery({ group_by, store_id })}`,
  )
}

export interface MovementSummaryRow {
  movement_type: string
  quantity: string
  movement_count: number
}

export function getInventoryMovements(params: ReportDateRange = {}): Promise<MovementSummaryRow[]> {
  return apiFetch<MovementSummaryRow[]>(`/api/v1/reports/inventory/movements${buildQuery(params)}`)
}

export interface ShrinkageRow {
  store_id: number
  reason_code: string
  quantity: string
  adjustment_count: number
}

export function getShrinkage(params: ReportDateRange = {}): Promise<ShrinkageRow[]> {
  return apiFetch<ShrinkageRow[]>(`/api/v1/reports/inventory/shrinkage${buildQuery(params)}`)
}

export interface TurnoverRow {
  product_id: number
  product_name: string
  cogs: string
  average_inventory_value: string | null
  turnover: string | null
  days_on_hand: string | null
}

export function getInventoryTurnover(
  date_from: string,
  date_to: string,
  store_id?: number[],
): Promise<TurnoverRow[]> {
  return apiFetch<TurnoverRow[]>(
    `/api/v1/reports/inventory/turnover${buildQuery({ date_from, date_to, store_id })}`,
  )
}

export interface SlowMovingRow {
  product_id: number
  product_name: string
  store_id: number
  quantity_on_hand: string
  units_sold_in_window: string
}

export function getSlowMoving(window_days = 90, store_id?: number[]): Promise<SlowMovingRow[]> {
  return apiFetch<SlowMovingRow[]>(
    `/api/v1/reports/inventory/slow-moving${buildQuery({ window_days, store_id })}`,
  )
}

export interface StockoutRow {
  product_id: number
  product_name: string
  store_id: number
}

export function getStockouts(store_id?: number[]): Promise<StockoutRow[]> {
  return apiFetch<StockoutRow[]>(`/api/v1/reports/inventory/stockouts${buildQuery({ store_id })}`)
}

export interface NegativeStockRow {
  product_id: number
  product_name: string
  store_id: number
  quantity_on_hand: string
}

export function getNegativeStock(store_id?: number[]): Promise<NegativeStockRow[]> {
  return apiFetch<NegativeStockRow[]>(
    `/api/v1/reports/inventory/negative-stock${buildQuery({ store_id })}`,
  )
}

export interface InTransitRow {
  transfer_id: number
  from_store_id: number
  to_store_id: number
  source_product_id: number
  destination_product_id: number
  quantity_in_transit: string
  value_in_transit: string
}

export function getInTransit(store_id?: number[]): Promise<InTransitRow[]> {
  return apiFetch<InTransitRow[]>(`/api/v1/reports/inventory/in-transit${buildQuery({ store_id })}`)
}

export interface GlReconciliationRow {
  label: string
  gl_balance: string
  operational_value: string
  discrepancy: string
}

export function getInTransitReconciliation(): Promise<GlReconciliationRow> {
  return apiFetch<GlReconciliationRow>('/api/v1/reports/inventory/in-transit-reconciliation')
}

export interface StockCountVarianceRow {
  stock_count_id: number
  product_id: number
  expected_quantity: string
  counted_quantity: string
  variance: string
}

export function getStockCountVariance(stockCountId: number): Promise<StockCountVarianceRow[]> {
  return apiFetch<StockCountVarianceRow[]>(
    `/api/v1/reports/inventory/stock-counts/${stockCountId}/variance`,
  )
}

// --- Purchasing & supplier analytics --------------------------------------

export interface PurchaseSpendRow {
  key: number
  label: string
  quantity_received: string
  spend: string
}

export function getPurchasingSpend(
  group_by: 'supplier' | 'store' | 'product',
  params: ReportDateRange = {},
): Promise<PurchaseSpendRow[]> {
  return apiFetch<PurchaseSpendRow[]>(
    `/api/v1/reports/purchasing/spend${buildQuery({ group_by, ...params })}`,
  )
}

export interface PoFulfillmentRow {
  purchase_order_id: number
  purchase_number: string
  status: string
  quantity_ordered: string
  quantity_received: string
  quantity_outstanding: string
}

export function getPoFulfillment(store_id?: number[]): Promise<PoFulfillmentRow[]> {
  return apiFetch<PoFulfillmentRow[]>(
    `/api/v1/reports/purchasing/po-fulfillment${buildQuery({ store_id })}`,
  )
}

export interface SupplierDeliveryPerformanceRow {
  supplier_id: number
  supplier_name: string
  receipt_count: number
  average_lead_time_days: string | null
}

export function getDeliveryPerformance(
  store_id?: number[],
): Promise<SupplierDeliveryPerformanceRow[]> {
  return apiFetch<SupplierDeliveryPerformanceRow[]>(
    `/api/v1/reports/purchasing/delivery-performance${buildQuery({ store_id })}`,
  )
}

export interface PurchasePriceVarianceRow {
  product_id: number | null
  total_variance: string
}

export function getPurchasePriceVariance(
  params: ReportDateRange = {},
): Promise<PurchasePriceVarianceRow[]> {
  return apiFetch<PurchasePriceVarianceRow[]>(
    `/api/v1/reports/purchasing/price-variance${buildQuery(params)}`,
  )
}

// --- Labor & payroll analytics --------------------------------------------

export interface HeadcountRow {
  store_id: number
  active_employee_count: number
}

export function getHeadcount(as_of?: string, store_id?: number[]): Promise<HeadcountRow[]> {
  return apiFetch<HeadcountRow[]>(
    `/api/v1/reports/payroll/headcount${buildQuery({ as_of, store_id })}`,
  )
}

export interface PayrollCostSummary {
  store_ids: number[] | null
  period_start: string
  period_end: string
  gross_pay: string
  deductions: string
  net_pay: string
  employer_contributions: string
  labor_cost: string
  regular_hours: string
  overtime_hours: string
  pending_period_count: number
  statutory_disclaimer: string
}

export function getPayrollCostSummary(
  period_start: string,
  period_end: string,
  store_id?: number[],
): Promise<PayrollCostSummary> {
  return apiFetch<PayrollCostSummary>(
    `/api/v1/reports/payroll/cost-summary${buildQuery({ period_start, period_end, store_id })}`,
  )
}

export interface LaborCostPercentRow {
  labor_cost: string
  net_sales: string
  labor_cost_percent: string | null
}

export function getLaborCostPercent(
  period_start: string,
  period_end: string,
  store_id?: number[],
): Promise<LaborCostPercentRow> {
  return apiFetch<LaborCostPercentRow>(
    `/api/v1/reports/payroll/labor-cost-percent${buildQuery({ period_start, period_end, store_id })}`,
  )
}

export function getPayrollReconciliation(store_id?: number[]): Promise<GlReconciliationRow> {
  return apiFetch<GlReconciliationRow>(
    `/api/v1/reports/payroll/reconciliation${buildQuery({ store_id })}`,
  )
}

// --- KPI dashboard ---------------------------------------------------------

export interface KpiDashboard {
  store_ids: number[] | null
  period_start: string
  period_end: string
  net_sales: string
  gross_margin_percent: string | null
  cogs_percent_of_net_sales: string | null
  average_transaction_value: string | null
  inventory_value: string
  inventory_turnover_note: string
  stockout_count: number
  shrinkage_units: string
  ap_outstanding: string
  ap_overdue: string
  purchase_spend: string
  labor_cost: string
  labor_cost_percent: string | null
  pending_payroll_periods: number
}

export function getKpiDashboard(
  period_start: string,
  period_end: string,
  store_id?: number[],
): Promise<KpiDashboard> {
  return apiFetch<KpiDashboard>(
    `/api/v1/reports/dashboard${buildQuery({ period_start, period_end, store_id })}`,
  )
}
